### PR TITLE
Add wpcom shortcodes

### LINKED ISSUE #5518: Add WordPress.com Shortcodes to Jetpack
Synopsis: We need to move some additional shortcodes from WordPress.com over into Jetpack so that users can more easily move from WordPress.com to WordPress.org

- [x] spotify
- [ ] ~~code (sourcecode, source)~~
- [x] tweet
- [x] googleapps
- [x] brightcove
- [ ] ~~playlist~~
- [ ] ~~display-posts~~
- [x] getty
- [ ] blog_subscription_form (needs parity with jetpacksubscriptionform)
- [ ] ~~list-pages (sibling-pages, child-pages)~~
- [ ] ~~ebay~~
- [ ] ~~vodpod~~
- [ ] ~~8tracks~~
- [x] archiveorg
- [x] gravatar
- [x] hulu
- [x] kickstarter
- [x] sitemap
- [x] lytro
- [x] mailchimp
- [x] quiz
- [ ] ~~upcomingevents~~
- [x] ustream

Responsible: Tentatively @eliorivero 

User experience: Functionality should be the same as on WordPress.com, see https://en.support.wordpress.com/shortcodes/

Complete: 0%

Difficulty:  23

Related: 87-gh-io

Last Updated: 2016-11-07 (by @jeherve)

### PR DESCRIPTION
**Pull request by eliorivero**
Fixes #5518

#### Changes proposed in this Pull Request:
Introduce new shortcodes in Jetpack

#### Testing instructions:
* run phpunit tests
* test shortcodes in editor

#### Proposed changelog entry for your changes:
Shortcodes: new are 

([see this comment for example shortcodes](https://github.com/Automattic/jetpack/pull/5838#issuecomment-266050670))

- [x] [spotify]
- [x] [tweet]
- [x] [googleapps]
- [x] [brightcove]
- [x] [getty]
- [x] [archiveorg]
- [x] [hulu]
- [x] [kickstarter]
- [x] [gravatar]
- [x] [gravatar_profile]
- [x] [quiz]
- [x] [sitemap]
- [x] [lytro]
- [x] [mailchip_subscriber_popup]
- [x] [ustream]
- [x] [ustreamsocial]

-------------------
- [x] Make sure your changes respect [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- [x] Did you make changes, or create a **new .js file**? If [Gulp](http://gulpjs.com/) is installed on your testing environment, run `gulp js:hint` before to commit your changes. It will allow you to [detect errors in Javascript files](http://jshint.com/about/).
- [x] Did you create a **new action or filter**? [Add inline documentation](https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/php/#4-hooks-actions-and-filters) to help others understand how to use the action or the filter.
- [x] Create **[unit tests](https://github.com/Automattic/jetpack/tree/master/tests)** if you can. If you're not familiar with Unit Testing, you can check [this tutorial](https://pippinsplugins.com/series/unit-tests-wordpress-plugins/).

### CODE REVIEW

**Comment by dereksmart** [comment]
For easy copy-pasta: Here are testible shortcodes.  

Archive Org
[archiveorg-book goodytwoshoes00newyiala]
[archiveorg-book http://www.archive.org/stream/goodytwoshoes00newyiala]
[archiveorg id=goodytwoshoes00newyiala width=480 height=430]
[archiveorg Experime1940]
[archiveorg http://archive.org/details/Experime1940 poster=http://archive.org/images/map.png]
[archiveorg id=Experime1940 width=640 height=480 autoplay=1]

Brightcove
[brightcove video_id="12345" account_id="99999"] 
[brightcove exp=627045696&vid=1415670151]
[brightcove exp=1463233149&vref=1601200825]

Getty 
[getty src="82278805" width="$width" height="$height"]

Google Apps

Gravatar
[gravatar email="user@example.org" size="48"]
[gravatar_profile who="user@example.org"]

Hulu
[hulu 369061]
[hulu id=369061]
[hulu id=369061 width=512 height=288 start_time="10" end_time="20" thumbnail_frame="10"]
[hulu http://www.hulu.com/watch/369061]
[hulu id=gQ6Z0I990IWv_VFQI2J7Eg width=512 height=288]

Kickstarter 
[kickstarter url="https://www.kickstarter.com/projects/peaktoplateau/yak-wool-baselayers-from-tibet-to-the-world" width="480" height=""]

Lytro.com
[lytro photo='202' show_arrow='true' show_border='true' show_first_time_user='true' allow_full_view='true']
[lytro username='lytroweb' photo='431119']

MailChimp
[mailchimp_subscriber_popup baseUrl="mc.us11.list-manage.com" uuid="1ca7856462585a934b8674c71" lid="2d24f1898b"]

Quiz
[quiz]
[question]What's the right answer?[/question]
[wrong]This one?[explanation]Nope[/explanation][/wrong]
[answer]Yes, this is the one![explanation]Yay![/explanation][/answer]
[wrong]Maybe this one[explanation]Keep trying[/explanation][/wrong]
[wrong]How about this one?[explanation]Try again[/explanation][/wrong]
[/quiz]

Sitemap
[sitemap]

Spotify
[spotify id="spotify:track:4bz7uB4edifWKJXSDxwHcs" width="400" height="100"]

Tweet
[tweet https://twitter.com/jack/statuses/20 width="350"]
[tweet tweet="https://twitter.com/jack/statuses/20" align="left" width="350" align="center" lang="es"]

Ustream
[ustream id=1524 live=1]
[ustreamsocial id=12980237 width="500"]

**Comment by samhotchkiss** [comment]
Gravatar hovercards is interacting with the gravatar shortcode:

gravatar-hovercard-fail

Tested Lytro, Getty, Archive.org and Sitemap so far, all look good!

**Comment by dereksmart** [comment]
I've tested these and all seem to be working as well as they do (if not better) on dotcom.  Nice work @eliorivero !

File: modules/shortcodes/brightcove.php [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+	/**
+	 * Convert to latest player format.
+	 *
+	 * @since 4.5.0
+	 *
+	 * @param array $atts Shortcode parameters.
+	 *
+	 * @return string
+	 */
+    static public function convert_to_new_studio( $atts ) {
```
Comment (dereksmart) [inline modules/shortcodes/brightcove.php]: Looks like this function is all indented with spaces

File: modules/shortcodes/getty.php [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+ * Compose shortcode based on Getty iframes.
+ *
+ * @since 4.5.0
+ *
+ * @param string $content
+ *
+ * @return mixed
+ */
+function wpcom_shortcodereverse_getty( $content ) {
+	if ( ! is_string( $content ) || false === stripos( $content, 'embed.gettyimages.com/embed' ) )
```
Comment (dereksmart) [inline modules/shortcodes/getty.php]: Should use brackets

File: modules/shortcodes/getty.php [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+		return $content;
+
+	$regexp = '!<iframe\s+src=[\'"](https?:)?//embed\.gettyimages\.com/embed(/|/?\?assets=)(\d+(,\d+)*)[^\'"]*?[\'"]((?:\s+\w+=[\'"][^\'"]*[\'"])*)((?:[\s\w]*))></iframe>!i';
+	$regexp_ent = str_replace( '&amp;#0*58;', '&amp;#0*58;|&#0*58;', htmlspecialchars( $regexp, ENT_NOQUOTES ) );
+
+	foreach ( array( 'regexp', 'regexp_ent' ) as $reg ) {
+		if ( !preg_match_all( $$reg, $content, $matches, PREG_SET_ORDER ) )
```
Comment (dereksmart) [inline modules/shortcodes/getty.php]: should use brackets

File: modules/shortcodes/getty.php [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+			if ( 'regexp_ent' == $reg )
+				$params = html_entity_decode( $params );
+
+			$params = wp_kses_hair( $params, array( 'http' ) );
+
+			$width = isset( $params['width'] ) ? (int) $params['width']['value'] : 0;
+			$height = isset( $params['height'] ) ? (int) $params['height']['value'] : 0;
+
+			$shortcode = '[getty src="' . esc_attr( $ids ) . '"';
+			if ( $width )
```
Comment (dereksmart) [inline modules/shortcodes/getty.php]: brackets

File: modules/shortcodes/getty.php [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+				$shortcode .= ' width="' . esc_attr( $width ) . '"';
+			if ( $height )
```
Comment (dereksmart) [inline modules/shortcodes/getty.php]: brackets

File: modules/shortcodes/getty.php [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+			$content = str_replace( $match[0], $shortcode, $content );
+		}
+	}
+
+	// strip out enclosing div and any other markup
+	$regexp = '%<div class="getty\s[^>]*+>.*?<div[^>]*+>(\[getty[^\]]*+\])\s*</div>.*?</div>%is';
+	$regexp_ent = str_replace( array( '&amp;#0*58;', '[^&gt;]' ), array( '&amp;#0*58;|&#0*58;', '[^&]' ), htmlspecialchars( $regexp, ENT_NOQUOTES ) );
+
+	foreach ( array( 'regexp', 'regexp_ent' ) as $reg ) {
+		if ( !preg_match_all( $$reg, $content, $matches, PREG_SET_ORDER ) )
```
Comment (dereksmart) [inline modules/shortcodes/getty.php]: brktz, missing space after `!`

File: modules/shortcodes/getty.php [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+ * @since 4.5.0
+ *
+ * @param array  $atts    Shortcode parameters.
+ * @param string $content Content enclosed by shortcode tags.
+ *
+ * @return string
+ */
+function jetpack_getty_shortcode( $atts, $content = '' ) {
+
+	if ( ! empty( $content ) )
```
Comment (dereksmart) [inline modules/shortcodes/getty.php]: Brackets throughout this whole conditional

File: modules/shortcodes/getty.php (hunk 3 above)
Comment (eliorivero) [inline modules/shortcodes/getty.php]: Where?

**Comment by eliorivero** [comment]
These are fixed now @dereksmart thanks for reviewing.

**Pull request merged**